### PR TITLE
chore(sql): fix subqueries returning a varchar column

### PR DESCRIPTION
**Pull request by jerrinot**
Some SQLs with subqueries returning a varchar column were broken. 

Example: 
```sql
select * from outer where sym in (select varchar from sub where symbol = 'baz'),
```
The IN operator accepts a subquery returning a varchar column and this was broken. 

Implementation note:
This bug was caused by some (delegating) `RecordCursorFactories` that expected inner `Record` to implicitly cast `VARCHAR` to `STRING` or `SYMBOL`. While `Function` performs such casting, `Record` does not. To complicate matters further, `VirtualRecord` appears to perform casting because it merely delegates to a function, which then handles the casting. This PR updates the affected `CursorRecordFactories` to explicitly convert `VARCHAR` columns to UTF-16 when necessary.

### CODE REVIEW

**Comment by jerrinot** [comment]
there are more cases to fix

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 1 / 1 (100.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/SqlCodeGenerator.java|1|1|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 24 / 28 (85.71%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cairo/sql/Record.java|7|11|63.64%|
|:large_blue_circle:|io/questdb/griffin/engine/table/FilterOnSubQueryRecordCursorFactory.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/table/LatestBySubQueryRecordCursorFactory.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/bool/InSymbolCursorFunctionFactory.java|12|12|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlCodeGenerator.java|1|1|100.00%|

**Comment by jerrinot** [comment]
it's ... not great. but doing this properly would require major changes. hopefully it will get simpler again once we start storing symbols as UTF8.

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 39 / 43 (90.70%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cairo/sql/Record.java|7|11|63.64%|
|:large_blue_circle:|io/questdb/griffin/engine/table/FilterOnSubQueryRecordCursorFactory.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/table/LatestBySubQueryRecordCursorFactory.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/bool/InSymbolCursorFunctionFactory.java|14|14|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlCodeGenerator.java|12|12|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 39 / 43 (90.70%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cairo/sql/Record.java|7|11|63.64%|
|:large_blue_circle:|io/questdb/griffin/engine/table/FilterOnSubQueryRecordCursorFactory.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/table/LatestBySubQueryRecordCursorFactory.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/bool/InSymbolCursorFunctionFactory.java|14|14|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlCodeGenerator.java|12|12|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 39 / 43 (90.70%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cairo/sql/Record.java|7|11|63.64%|
|:large_blue_circle:|io/questdb/griffin/engine/table/FilterOnSubQueryRecordCursorFactory.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/table/LatestBySubQueryRecordCursorFactory.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/bool/InSymbolCursorFunctionFactory.java|14|14|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlCodeGenerator.java|12|12|100.00%|

File: core/src/main/java/io/questdb/griffin/engine/functions/bool/InSymbolCursorFunctionFactory.java [hunk 1]
```diff
@@ -172,7 +185,7 @@ public void toPlan(PlanSink sink) {
         private void buildValueSet() {
             final Record record = cursor.getRecord();
             while (cursor.hasNext()) {
-                CharSequence value = func.get(record, 0);
+                CharSequence value = func.get(record, 0, sink);
```
Comment (ideoma) [inline core/src/main/java/io/questdb/griffin/engine/functions/bool/InSymbolCursorFunctionFactory.java:188]: You can use `Misc.getThreadLocalSink()` here instead of the dedicated sink. Thread static getter will be used once, relatively not expensive.

File: core/src/main/java/io/questdb/griffin/engine/functions/bool/InSymbolCursorFunctionFactory.java [hunk 2]
```diff
@@ -282,7 +297,7 @@ private void buildSymbolKeys() {
 
             final Record record = cursor.getRecord();
             while (cursor.hasNext()) {
-                int key = symbolTable.keyOf(func.get(record, 0));
+                int key = symbolTable.keyOf(func.get(record, 0, sink));
```
Comment (ideoma) [inline core/src/main/java/io/questdb/griffin/engine/functions/bool/InSymbolCursorFunctionFactory.java:300]: You can use `Misc.getThreadLocalSink()` here

File: core/src/main/java/io/questdb/griffin/engine/table/FilterOnSubQueryRecordCursorFactory.java [hunk 3]
```diff
@@ -221,7 +224,7 @@ private void buildFactories() {
             final StaticSymbolTable symbolTable = delegate.getDataFrameCursor().getSymbolTable(columnIndex);
             final Record record = baseCursor.getRecord();
             while (baseCursor.hasNext()) {
-                final CharSequence symbol = func.get(record, 0);
+                final CharSequence symbol = func.get(record, 0, sink);
```
Comment (ideoma) [inline core/src/main/java/io/questdb/griffin/engine/table/FilterOnSubQueryRecordCursorFactory.java:227]: You can use Misc.getThreadLocalSink() here

File: core/src/main/java/io/questdb/griffin/engine/table/LatestBySubQueryRecordCursorFactory.java [hunk 4]
```diff
@@ -220,7 +223,7 @@ private void buildSymbolKeys() {
             final StaticSymbolTable symbolTable = delegate.getSymbolTable(columnIndex);
             final Record record = baseCursor.getRecord();
             while (baseCursor.hasNext()) {
-                int symbolKey = symbolTable.keyOf(func.get(record, 0));
+                int symbolKey = symbolTable.keyOf(func.get(record, 0, sink));
```
Comment (ideoma) [inline core/src/main/java/io/questdb/griffin/engine/table/LatestBySubQueryRecordCursorFactory.java:226]: You can use Misc.getThreadLocalSink() here

File: core/src/main/java/io/questdb/griffin/engine/functions/bool/InSymbolCursorFunctionFactory.java (hunk 1 above)
Comment (jerrinot) [reply core/src/main/java/io/questdb/griffin/engine/functions/bool/InSymbolCursorFunctionFactory.java:188]: I was considering this. but then I decided it was asking for troubles: we don't know what `func.get()` will do. if it's an actual physical record then it's safe. but it can be a virtual record backed by an arbitrary function. and if that function uses a thread-local sink too then things will break in a spectacular (and hard to analyze) way. 

wdyt?

File: core/src/main/java/io/questdb/cairo/sql/Record.java [hunk 5]
```diff
@@ -460,8 +471,9 @@ interface CharSequenceFunction {
         /**
          * @param record to retrieve CharSequence from
          * @param col    numeric index of the column
+         * @param sink   sink the function can use if a conversion is required
          * @return record as a char sequence
          */
-        CharSequence get(Record record, int col);
+        CharSequence get(Record record, int col, DirectUtf16Sink sink);
```
[the remark this answers is not in the document]
Comment (jerrinot) [reply core/src/main/java/io/questdb/cairo/sql/Record.java:477]: originally I did have `StringSink`. it was simpler - no lifecycle management. but then I decided not to allocate on Java heaps. 
happy to change it back if you think it's alright.

**Review by ideoma — APPROVED** [review]

**Comment by bluestreak01** [comment]
@jerrinot could you rephrase PR title and add description. The existing title is very cryptic.

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 35 / 39 (89.74%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cairo/sql/Record.java|7|11|63.64%|
|:large_blue_circle:|io/questdb/griffin/engine/table/FilterOnSubQueryRecordCursorFactory.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/table/LatestBySubQueryRecordCursorFactory.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/bool/InSymbolCursorFunctionFactory.java|12|12|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlCodeGenerator.java|12|12|100.00%|
